sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala

```diff
@@ -280,61 +295,60 @@ class GpuDynamicPartitionDataSingleWriter(
     taskAttemptContext: TaskAttemptContext,
     committer: FileCommitProtocol)
   extends GpuFileFormatDataWriter(description, taskAttemptContext, committer) {
+  /** Wrapper class to index a unique concurrent output writer. */
+  protected case class WriterIndex(
+      var partitionPath: Option[String],
+      var bucketId: Option[Int]) {
+
+    override def hashCode(): Int = ScalaMurmur3Hash.productHash(this)
+
+    override def equals(obj: Any): Boolean = {
+      if (this.eq(obj.asInstanceOf[AnyRef])) {
+        true
+      } else if (obj.isInstanceOf[WriterIndex]) {
+        obj.hashCode() == this.hashCode()
```

Comment:
I am not okay with this. I know it should be super rare that two WriterIndex objects would have a hash collision, but it is possible. I am also not happy that a WriterIndex is a case class, but it is mutable.  That violates a lot of the tenants of what a case class is supposed to be and so the generated code, for things like equals and hashCode don't work, which is why you are overwriting them.  Can we please make this a regular class instead?

Reply:
Really appreciate for catching this.  I can not remeber what was in my mind when writting this piece of code.